sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_command_buffer.cpp

```diff
@@ -0,0 +1,777 @@
+  const ze_copy_region_t ZeDstRegion = {DstOriginX, DstOriginY, DstOriginZ,
+                                        Width,      Height,     Depth};
+
+  std::vector<ze_event_handle_t> ZeEventList;
+  ur_result_t Res = getEventsFromSyncPoints(CommandBuffer->SyncPoints,
+                                            NumSyncPointsInWaitList,
+                                            SyncPointWaitList, ZeEventList);
+
+  ur_event_handle_t LaunchEvent;
+  Res = EventCreate(CommandBuffer->Context, nullptr, true, &LaunchEvent);
+  if (Res)
+    return UR_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+  LaunchEvent->CommandType = CommandType;
+
+  ZE2UR_CALL(zeCommandListAppendMemoryCopyRegion,
+             (CommandBuffer->ZeCommandList, Dst, &ZeDstRegion, DstPitch,
+              DstSlicePitch, Src, &ZeSrcRegion, SrcPitch, SrcSlicePitch,
+              LaunchEvent->ZeEvent, ZeEventList.size(), ZeEventList.data()));
+
+  urPrint("calling zeCommandListAppendMemoryCopyRegion() with"
+          "  ZeEvent %#lx\n",
+          ur_cast<std::uintptr_t>(LaunchEvent->ZeEvent));
+
+  // Get sync point and register the event with it.
+  *SyncPoint = CommandBuffer->GetNextSyncPoint();
+  CommandBuffer->RegisterSyncPoint(*SyncPoint, LaunchEvent);
+  return UR_RESULT_SUCCESS;
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferCreateExp(
+    ur_context_handle_t hContext, ur_device_handle_t hDevice,
+    const ur_exp_command_buffer_desc_t *pCommandBufferDesc,
+    ur_exp_command_buffer_handle_t *phCommandBuffer) {
+  // Force compute queue type for now. Copy engine types may be better suited
+  // for host to device copies.
+  uint32_t QueueGroupOrdinal =
+      hDevice
+          ->QueueGroup[ur_device_handle_t_::queue_group_info_t::type::Compute]
+          .ZeOrdinal;
+
+  ZeStruct<ze_command_list_desc_t> ZeCommandListDesc;
+  ZeCommandListDesc.commandQueueGroupOrdinal = QueueGroupOrdinal;
+
+  ze_command_list_handle_t ZeCommandList;
+  ZE2UR_CALL(zeCommandListCreate, (hContext->ZeContext, hDevice->ZeDevice,
+                                   &ZeCommandListDesc, &ZeCommandList));
+  try {
+    *phCommandBuffer = new ur_exp_command_buffer_handle_t_(
+        hContext, hDevice, ZeCommandList, ZeCommandListDesc,
+        pCommandBufferDesc);
+  } catch (const std::bad_alloc &) {
+    return UR_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+  } catch (...) {
+    return UR_RESULT_ERROR_UNKNOWN;
+  }
+
+  // Create signal & wait events to be used in the command-list for sync
+  // on command-buffer enqueue.
+  auto hCommandBuffer = *phCommandBuffer;
+  UR_CALL(EventCreate(hContext, nullptr, true, &hCommandBuffer->SignalEvent));
+  UR_CALL(EventCreate(hContext, nullptr, false, &hCommandBuffer->WaitEvent));
+
+  // Add prefix commands
+  ZE2UR_CALL(zeCommandListAppendEventReset,
+             (ZeCommandList, hCommandBuffer->SignalEvent->ZeEvent));
+  ZE2UR_CALL(zeCommandListAppendBarrier,
+             (ZeCommandList, nullptr, 1, &hCommandBuffer->WaitEvent->ZeEvent));
```

Comment:
so the first command in the command buffer is an eventReset, and the second is a barrier waiting for an event (signaled from somewhere else) that we need to wait for this command buffer to start, correct?